### PR TITLE
Add to analysis doc (from new branch)

### PR DESCRIPTION
**Pull request by milankl**
Duplicate of #504 but from a new branch #miniufo-main that is in SpeedyWeather.jl for the docs to deploy

### CODE REVIEW

File: docs/src/analysis.md [hunk 1]
```diff
@@ -108,46 +112,47 @@ which is _exactly_ the same. So mass is conserved, woohoo.
...
-total volume (times ``h`` for the vertical then integrated over the sphere``\iint dA``).
+The total energy in the [shallow water equations](@ref shallow_water_model) is the sum of the kinetic energy density
+``\frac{1}{2}(u^2 + v^2)`` and the potential energy density ``gz`` integrated over the
+total volume (times ``h=\eta+H-H_b`` for the vertical then integrated over the sphere
+``\iint dA``).
 
 ```math
-\iint \frac{h}{2}(u^2 + v^2) + gh^2 dA
+E = \iint \left[ \int_{H_b}^{\eta}\frac{1}{2}\left(u^2 + v^2 + gz\right)dz \right]dA \\
+  = \iint \frac{1}{2}\left(u^2 + v^2 + gh\right)h dA
```
Comment (navidcy) [inline docs/src/analysis.md:128]:
```suggestion
\begin{align}
E & = \iint \left[ \int_{H_b}^{\eta}\frac{1}{2}\left(u^2 + v^2 + gz\right)dz \right]dA \\
  & = \iint \frac{1}{2}\left(u^2 + v^2 + gh\right)h dA
\end{align}
```

File: docs/src/analysis.md [hunk 2]
```diff
@@ -108,46 +112,47 @@ which is _exactly_ the same. So mass is conserved, woohoo.
...
     Hb = model.orography.orography
     g = model.planet.gravity
     
-    @. h = η + H - Hb
-    @. E = h/2*(u^2 + v^2) + g*h^2
+    @. h = η + H - Hb  # layer thickness between the bottom and free surface
+    @. E = h/2*(u^2 + v^2) + g*h^2 # vertically-integrated mechanical energy
 
     # transform to spectral, take l=m=0 mode at [1] and normalize for mean
     E_mean = real(spectral(E)[1]) / model.spectral_transform.norm_sphere
```
Comment (navidcy) [inline docs/src/analysis.md:150]:
```suggestion
    return E_mean = real(spectral(E)[1]) / model.spectral_transform.norm_sphere
```

File: docs/src/analysis.md [hunk 3]
```diff
@@ -221,12 +237,374 @@ plot(q)
...
+and ``R`` the radius of Earth.
+
+Following previous examples, let us define a `total_angular_momentum` function as
+
+```@example analysis
+using SpeedyWeather
+
+function total_angular_momentum(u, η, model)
+    h = zero(u)
+    Λ = zero(u)                             # allocate grid variable
```
Comment (navidcy) [inline docs/src/analysis.md:258]:
```suggestion
    # allocate grid variable
    h = zero(u)
    Λ = zero(u)
```

File: docs/src/analysis.md [hunk 4]
```diff
@@ -221,12 +237,374 @@ plot(q)
...
+
+    # transform to spectral, take l=m=0 mode at [1] and normalize for mean
+    C_mean = real(spectral(ζ .+ f)[1]) / model.spectral_transform.norm_sphere
+end
+```
+
+!!! note "Global-integrated circulation"
+    Note that the area integral of relative vorticity ``\zeta`` and planetary vorticity
+    ``f`` over the whole surface of a sphere are analytically exactly zero.
+    Numerically, ``C_mean`` should be a small number but may not be exactly zero due
```
Comment (navidcy) [inline docs/src/analysis.md:315]:
```suggestion
    Numerically, `C_mean` should be a small number but may not be exactly zero due
```

File: docs/src/analysis.md [hunk 5]
```diff
@@ -221,12 +237,374 @@ plot(q)
...
 ## Circulation
 
-More to follow ...
+Total circulation is defined as the area-integrated absolute vorticity:
+
+```math
+C = \iint \left(\zeta + f\right) dA
+```
+
+Following previous fashion, we define a function ``total_circulation`` for this:
```
Comment (navidcy) [inline docs/src/analysis.md:301]:
```suggestion
Following previous fashion, we define a function `total_circulation` for this:
```

File: docs/src/analysis.md [hunk 6]
```diff
@@ -221,12 +237,374 @@ plot(q)
...
+
+```@example analysis
+function total_circulation(ζ, model)
+    f = coriolis(ζ)  # create f on that grid
+
+    # transform to spectral, take l=m=0 mode at [1] and normalize for mean
+    C_mean = real(spectral(ζ .+ f)[1]) / model.spectral_transform.norm_sphere
```
Comment (navidcy) [inline docs/src/analysis.md:308]:
```suggestion
    return C_mean = real(spectral(ζ .+ f)[1]) / model.spectral_transform.norm_sphere
```

File: docs/src/analysis.md [hunk 7]
```diff
@@ -221,12 +237,374 @@ plot(q)
...
-## Enstrophy
 
-More to follow ...
\ No newline at end of file
+## Online diagnostics
+
+Now we want to calculate all the above global diagnostics periodically
+during a simulation. For that we will use [Callbacks](@ref), which
+let us inject code into a simulation that is executed after every
+time step (or at any other scheduled time, see [Schedules]@ref)).
```
Comment (navidcy) [inline docs/src/analysis.md:374]:
```suggestion
time step (or at any other scheduled time, see [`Schedules`]@ref)).
```

File: docs/src/analysis.md [hunk 8]
```diff
@@ -221,12 +237,374 @@ plot(q)
...
+    
+    callback.time[1] = progn.clock.time
+    callback.M[1] = M  # set initial conditions
+    callback.C[1] = C  # set initial conditions
+    callback.Λ[1] = Λ  # set initial conditions
+    callback.K[1] = K  # set initial conditions
+    callback.P[1] = P  # set initial conditions
+    callback.Q[1] = Q  # set initial conditions
+    
+    callback.timestep_counter = 1  # (re)set counter to 1
```
Comment (navidcy) [inline docs/src/analysis.md:498]:
```suggestion
    callback.timestep_counter = 1  # (re)set counter to 1
    
    return nothing
```

File: docs/src/analysis.md [hunk 9]
```diff
@@ -221,12 +237,374 @@ plot(q)
...
+    defDim(ds, "time", n_timesteps)
+    defVar(ds, "time",                  callback.time,  ("time",))
+    defVar(ds, "mass",                  callback.M,     ("time",))
+    defVar(ds, "circulation",           callback.C,     ("time",))
+    defVar(ds, "angular momentum",      callback.Λ,     ("time",))
+    defVar(ds, "kinetic energy",        callback.K,     ("time",))
+    defVar(ds, "potential energy",      callback.P,     ("time",))
+    defVar(ds, "potential enstrophy",   callback.Q,     ("time",))
+    
+    close(ds)
```
Comment (navidcy) [inline docs/src/analysis.md:545]:
```suggestion
    close(ds)

    return nothing
```

File: docs/src/analysis.md [hunk 10]
```diff
@@ -221,12 +237,374 @@ plot(q)
...
+function total_angular_momentum(u, η, model)
+    h = zero(u)
+    Λ = zero(u)                             # allocate grid variable
+    
+    H = model.atmosphere.layer_thickness
+    Hb = model.orography.orography
+    R = model.spectral_grid.radius
+    Ω = model.planet.rotation
+
+    r = R * cos.(model.geometry.lats)       # create r on that grid
```
[the remark this answers is not in the document]
Comment (navidcy) [reply docs/src/analysis.md:269]: great then!

**Comment by milankl** [comment]
@miniufo this looks great now, I'm merging!! 🎉

**Pull request merged**